### PR TITLE
Reworked the porting guide, adding new material without detracting from what was already there.

### PR DESCRIPTION
**Pull request by abkro**

### CODE REVIEW

**Review by doniks — CHANGES_REQUESTED** [review]
Hey Ari. Thanks for this PR!

I've reviewed sections 1, 2 and 3. There are some detailed points inline. From a bigger picture, I would suggest to reorder a few sections (comments/reasons in brackets):

Intro

Build: (merge 3.2 and 3.3?)
- fix mounts 
- kernel config
- build images


Install:
- flash halium-boot
- install system.img and rootfs
- ssh 

Bring up Unity8:
- udev (udev has to happen sooner than other things because it is more blocking, use RW rootfs to keep it simple)
- lxc/mir debugging

Configure UT:
- overlay config (fix udev and prep for subsequent steps)
- Display settings (moved down so it is after lxc and overlay)
- apparmor (moved down after lxc and the simpler and more obvious display cfg)
- sound (after lxc and apparmor, where one is more likely to have apps that produce sound)


Furthermore, wrt to section layout, I am wondering about the following points. I'm completely settled on these, but I'm just thinking out loud/asking your opinion: 

* the 7.1 page is VERY long, maybe breaking it into smaller pages per subsection? but I'm not sure
* the current guide has the common errors right after to steps where they are most likely to occur. As a porter I want to see the typical ssh gotchas at the moment when I try to establish ssh connection. Similarly mir/lxc problems are relevant when I am at that step. If one would move the errors back to be subsequent to the section where one produces the errors, then the page length also becomes much longer yet and the splitting up becomes more interesting ...

File: porting/3_Halium_7-1_porting.rst [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+1. ``cd BUILDDIR``
+2. ``source build/envsetup.sh``
+3. ``breakfast devicename``
+4. ``mka halium-boot``
+5. ``mka systemimage``
+
+.. _3.3.4:
+
+3.3.4   Apply apparmor kernel patches
```
[the remark this answers is not in the document]
Comment (abkro) [reply porting/3_Halium_7-1_porting.rst:70]: That is why the section starts with the sentence: "At this point you kan skip to the next section and test installing your build, bla bla." 
Not sure you noticed this, so I will leave it for now...

File: porting/3_Halium_7-1_porting.rst [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+See section :ref:`4.2 <4.2>`
+
+.. _3.5:
+
+3.5 Bring up Unity 8 (the graphical UI)
+---------------------------------------
+
+Now that you're logged in, there are a few more steps before Ubuntu Touch will be fully functional on your device.
+
+3.5.1   Make / writable
```
[the remark this answers is not in the document]
Comment (abkro) [reply porting/3_Halium_7-1_porting.rst:164]: I changed the wording to something resembling what you suggested here.

Wrt the section layout and the page length, this was one of the reasons why I placed the errors/issues in a separate file. Not all porterts will experience all the errors, and so, linking them from the section in question will make them easily accessible while permitting those who don't need this info to proceed more easily.

But the page remains very long, as you have pointed out. I am now going to try to divide it up into separate pages that reflect the part in the intro where the phases of the porting process are described. That should make the whole guide more consistent and easier to follow, I believe.

File: porting/1_Intro.rst [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+.. _1.5:
+
+1.5 Getting community help
+--------------------------
+
+When you run into trouble, and you will, refer to one or more of the sources below:
+
+* IRC: #halium on Freenode 
+* Matrix: #halium:matrix.org 
+* Telegram: @halium
```
[the remark this answers is not in the document]
Comment (abkro) [reply porting/1_Intro.rst:91]: I know this sounds kind of helpless, but don't know how to find the link.

File: porting/1_Intro.rst [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+Luckily, there is a community of porters out there who are eager to see Ubuntu Touch ported to new devices. When you run into trouble, you should search the sources below (Getting help) to see if others have before you have solved the issue. There are also online Telegram chat groups you can join to ask for help, but please bear in mind that those participating are doing so of their own free will, and in their spare time.
+
+.. _1.4:
+
+1.4 Prior knowledge and skills
+------------------------------
+
+Porters come in all sizes and shapes, so to speak. Therefore, this guide does not presuppose extensive knowledge or skills in any particular field. You should, however, as a bare minimum be familiar with some common shell commands and be comfortable working from the terminal on you host pc. Furthermore, the guide is based on a host PC running Linux. If you have some knowledge of programming, this will come in handy at some point, especially if you are familiar with C / C++. Also, you should familiarize yourself with git, setup a github account and start documenting your steps as described in this guide and the Halium porting guide from the very beginning.
```
[the remark this answers is not in the document]
Comment (abkro) [reply porting/1_Intro.rst:78]: Done

File: porting/1_Intro.rst [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+Thus an Ubuntu Touch port is composed of the these components:
+    * The Ubuntu Touch (UT) root filesystem (rootfs)
+    * `Halium <https://halium.org/>`_ (contained in the boot and system images)
+    * The vendor blobs
+
+You, the porter, need to build Halium (i.e. halium-boot.img and system.img) and install these together with the UBports Ubuntu Touch rootfs in order to create a functioning Ubuntu Touch port. 
+
+Halium is built using source code for a modified version of the Android operating system called LineageOS. The required source code is available online and needs to be downloaded and configured to build the correct Halium version for each individual device port. Table 1 shows which versions are required for the different Halium versions.
+
+Table 1
```
Comment (doniks) [inline porting/1_Intro.rst:34]: I think we could get rid of the heading for this table. It shows up as a section header in the navigation. 

Simply say "The table below" or similar

File: porting/1_Intro.rst [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+This is why Ubuntu Touch cannot be built as a complete standalone OS for most commercial devices. Instead, porting the system to these commercially available devices involves integrating these low level components into the rest of the system which can be built from source.
+
+The next component of Ubuntu Touch is a pre-compiled root filesystem which needs to be installed on the device. This component is not capable of communicating directly with the device hardware through the above-mentioned device drivers. Instead, this communication is mediated by a Hardware Abstraction Layer (HAL) which needs to be built for each specific device, because each device has its specific hardware architecture. This component is called Halium and is available in different versions (5.1 which is largely obsolete, 7.1, and 9 as of writing) corresponding to different Android versions.
```
[the remark this answers is not in the document]
Comment (doniks) [reply porting/1_Intro.rst:19]: current draft still says not capable ...

File: porting/1_Intro.rst [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+Halium is an indispensible part of an Ubuntu Touch port and is available in the form of open source software. Developing a new version of Halium is a very considerable task which is the why only a few versions of Halium are available. Each port of Ubuntu Touch has to be based on one of the available Halium versions and vendor blobs from the corresponding Android version. See the first two columns of Table 1 (below) for details.
+
+Thus an Ubuntu Touch port is composed of the these components:
+    * The Ubuntu Touch (UT) root filesystem (rootfs)
+    * `Halium <https://halium.org/>`_ (contained in the boot and system images)
+    * The vendor blobs
+
+You, the porter, need to build Halium (i.e. halium-boot.img and system.img) and install these together with the UBports Ubuntu Touch rootfs in order to create a functioning Ubuntu Touch port.
```
[the remark this answers is not in the document]
Comment (doniks) [reply porting/1_Intro.rst:30]: still

File: porting/1_Intro.rst [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+.. _1.2:
+
+1.2 Overview of the porting process
+-----------------------------------
+
+    * Preparations
+        In this first stage you will choose a target device, gather information about it and choose a suitable workstation (PC or laptop) on which to do your work.
+    * Set up your Halium build environment and complete a test build
+        This stage involves downloading all the relevant source code and securing the vendor blobs. You will then build the Halium boot image and system image according to the Halium guide to ensure that all your sources are set up correctly and work in unison. You will *not* be installing this build! You can view it as a "trial run".
```
[the remark this answers is not in the document]
Comment (doniks) [reply porting/1_Intro.rst:52]: mhm. it still says "you will *not"" ... we discussed previously, that in fact both of us *did* and that we found it helpful ... and it certainly doesn't harm

File: porting/1_Intro.rst [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+1.5 Getting community help
+--------------------------
+
+When you run into trouble, and you will, refer to one or more of the sources below:
+
+* IRC: #halium on Freenode 
+* Matrix: #halium:matrix.org 
+* Telegram: @halium
+
+Make sure to check the `UBports Forum <https://forums.ubports.com/category/33/porting>`_ as well.
```
[the remark this answers is not in the document]
Comment (doniks) [reply porting/1_Intro.rst]: https://t.me/ubports_porting

File: porting/2_Preparations.rst [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+.. _2.2:
+
+2.2 Locate relevant guides and other information
+------------------------------------------------
+
+If your desired target device complies to the above, the next thing to do is to locate the available guides and other documentation. This step will help you later on when you run into issues developing your port. For now, the main thing to look for is how to unlock your bootloader. (See 2.3, below).
+
+Head over to the `LineageOS Wiki <https://wiki.lineageos.org/>`_. Look up your device and read and bookmark the guides that are listed there. Try a web search for additional information, specifying both its retail name and the code name you found on the LineageOS site. Be careful to check that the information you gather applies to your specific device, keeping in mind that many devices are sold in a number of different variants with different hardware specifications.
+
+Another rich source of information is the `XDA Developer Forum <https://www.xda-developers.com/>`_.
```
[the remark this answers is not in the document]
Comment (doniks) [reply porting/2_Preparations.rst:19]: still

File: porting/2_Preparations.rst [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+.. _2.3:
+
+2.3 Unlock the bootloader
+-------------------------
+
+Vendors provide their devices in a locked bootloader state. This is a kind of software "seal" intended to prevent you from making modifications directly to the operative system and system software. If you unlock the bootloader, you will be able to make such modifications, but in this state your device's warranty will be void. The choice is yours, but this step is mandatory if you wish to install Ubuntu Touch on the device.
```
[the remark this answers is not in the document]
Comment (doniks) [reply porting/2_Preparations.rst]: still

[the remark this answers is not in the document]
Comment (doniks) [reply porting/2_Preparations.rst]: still

[the remark this answers is not in the document]
Comment (doniks) [reply porting/2_Preparations.rst]: still

File: porting/2_Preparations.rst [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+-------------------------
+
+You need to install a custom recovery image on your device in order to handle formatting and flashing image files onto it. Head over to the `TeamWin website <https://twrp.me/Devices/>`_  and locate the image file for your device. Follow the installation instructions provided on the website.
+
+.. _2.5:
+
+2.5 Recommendations for the host/build PC 
+-----------------------------------------
+
+The remainder of this guide presumes you are using a build pc running Linux. Although high performance always is nice, stability and sufficient ROM and harddisk space are the main concerns here. You do not need the latest and the best hardware, nor do you need the latest distribution release. In fact, it is not uncommon to run into issues when choosing the very newest release, as some of the software needed may not yet have been built for it. For example, as of writing the latest release of Ubuntu Linux is 20.04, but many still recommend using the previous long term support release, 18.04.
```
[the remark this answers is not in the document]
Comment (doniks) [reply porting/2_Preparations.rst]: still

File: porting/3_Halium_7-1_porting.rst [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+3.1 Introduction
+----------------
+
+A Halium 7.1 - Ubuntu Touch port consists of the following parts:
+    
+    * The boot image, halium-boot.img
+    * The system image, system.img
+    * The UBports - Ubuntu Touch - root file system (hereafter referred to as the *rootfs*)
+
+The boot image contains the Linux kernel, the heart of the operative system, which needs to be built from source code for each specific device. The system image is also specific to the device and is similarly built from source code. The rootfs, on the other hand, is maintained by the UBports Foundation and is downloadable in the form of a prebuilt .tar.gz bundle which is then installed together with the system.img.
```
[the remark this answers is not in the document]
Comment (doniks) [reply porting/3_Halium_7-1_porting.rst]: still

File: porting/3_Halium_7-1_porting.rst [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+
+.. _3.2:
+
+3.2 Build the Halium system image
+---------------------------------
+
+First you need to build the Halium system image according to the the Halium porting guide. This is not a build you will install. It is simply a preparation for the actual build for your port. In other words it a test to confirm that you have set up a viable source tree that can be used to do the final build.
```
[the remark this answers is not in the document]
Comment (doniks) [reply porting/3_Halium_7-1_porting.rst]: still

File: porting/3_Halium_7-1_porting.rst [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+
+Head over to `the Halium porting guide <http://docs.halium.org/en/latest/porting/first-steps.html#set-up-your-build-device>`_ and follow the steps described there remembering that you are building Halium 7.1.
+
+Once the system image has been successfully built (without errors), you then know that your source tree is correctly set up to proceed. You then return here and proceed with subsection 3.3 below.
```
[the remark this answers is not in the document]
Comment (doniks) [reply porting/3_Halium_7-1_porting.rst]: still

File: porting/1_Intro.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (doniks) [reply porting/1_Intro.rst:93]: :) it's not too obvious. in telegram linux desktop client I enter the chat, then click on the name of the group and get to "Group Info", there it is: https://t.me/ubports_porting

File: porting/3_Halium_7-1_porting.rst [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+3.3.4   Apply apparmor kernel patches
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+At this point you can also skip to section 3.2 and test installing your build. However, you will have to come back and complete the step in this section in order to complete a functional port.
+
+`What is apparmor? <https://wiki.ubuntu.com/AppArmor>`_
+
+Start by downloading the backported `apparmor patch <https://github.com/ubports/apparmor-backports-ut>`_ corresponding to your device's kernel version. Your kernel version is specified at the beginning of your kernel defconfig file, i.e. the one you edited in section :ref:`3.3.2 <3.3.2>` above.
+
+You now need to delete your entire BUILDDIR/kernel/VENDOR/MODEL/security/apparmor subdirectory and replace it with the one you downloaded. Then rebuild halium-boot.
```
[the remark this answers is not in the document]
Comment (doniks) [reply porting/3_Halium_7-1_porting.rst]: mhm. I see there is a ut repo of apparmor with explicit delete/replace steps ... I'm pretty sure I did this differently, but ok, there are certainly multiple ways to skin this cat. But that is out of scope of this PR. so: resolved

File: porting/3_Halium_7-1_porting.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (doniks) [reply porting/3_Halium_7-1_porting.rst]: mhm. maybe we were talking past each other here .... however, I see that now apparmor is towards the end. I think it is fine where it is now, so: resolved

File: porting/3_4_Halium_7-1_porting.rst [hunk 17]
```diff
@@ -0,0 +1,117 @@
+Configuring Ubuntu Touch (7.1)
+==============================
```
Comment (doniks) [inline porting/3_4_Halium_7-1_porting.rst:2]: this file has three heading layers:

Configuring UT
+ Configuring and eable remaining device functionality
++ Udev rules
++ Display scaling
++ Apply apparmor kernel patches
++ Sound configuration
+ Further configuration

I propose to move udev, scaling, apparmor and sound one level up and split the "Configure and enable remaining.." section into a) a general intro and a new "overlay" section. Like this:

Configuring UT 
"In the previous section you adjusted the udev settings .... overlay files, as described below."
+ Overlay files 
"Overlay files allow you to deploy configurations that are specific to your port in a place that normally belongs to the rootfs. Note that as a rule of thumb ... "
+ Udev rules
+ Display scaling
+ Apply apparmor kernel patches
+ Sound configuration
+ Further configuration

**Comment by doniks** [comment]
thanks for the rework. 

I have gone through the individual remarks as best as I could and mostly only rechecked locally. I think we're getting close to the finish. If you rework it once more, I think we should at least have the section structure nailed and most of the individual points resolved. At that point I would like to once more read the whole thing end to end to make sure it all still flows.  

I've made/updated a couple of inline remarks again.

Two general points:
1) it doesn't build right now, because it seems some sections are duplicated. I've locally removed from 3_2 what was also present in 3_3 and in 3_4 and reviewed that build
2) I like the separate files/sections much better than the previous draft. However, now all sections are top level. This means that on the navigation bar on the side, there are A LOT of items inside porting. I think this is overwhelming and it would be better to condense this. I guess we could do: 

* one single top section "porting" and stick everything inside, or
* five top sections: Intro, Halium 7, Halium 9, Recovery, Installer

File: porting/3_4_Halium_7-1_porting.rst (hunk 17 above)
Comment (abkro) [reply porting/3_4_Halium_7-1_porting.rst:2]: I agree. I wonder if the same should be done for two of the other files. Will try...

**Comment by abkro** [comment]
Point 1: This happened because I had forgotten to remove a bunch of things from one of the files when splitting them up.
Point 2: Yes, I saw this before pushing my changes, but chose to leave it that way to see what your opinion was. I'll find some way of changing it.
Point 3: I implemented this change already.

**Review by doniks — APPROVED** [review]

**Comment by doniks** [comment]
thanks a buuunch!!!!111

**Pull request merged**